libp2p/src/main/kotlin/io/libp2p/pubsub/gossip/GossipRouter.kt

```diff
@@ -565,16 +595,30 @@ open class GossipRouter(
         }
     }
 
+    private fun peerDoesNotWantMessage(peer: PeerHandler, messageId: MessageId): Boolean {
+        return peerIDontWant[peer]?.messageIdsAndTimeReceived?.contains(messageId) == true
+    }
+
     private fun iWant(peer: PeerHandler, messageIds: List<MessageId>) {
         if (messageIds.isEmpty()) return
         messageIds[random.nextInt(messageIds.size)]
             .also { iWantRequests[peer to it] = currentTimeSupplier() }
         enqueueIwant(peer, messageIds)
     }
 
+    private fun iDontWant(msg: PubsubMessage, receivedFrom: PeerHandler) {
+        if (!this.protocol.supportsIDontWant()) return
+        if (msg.protobufMessage.data.size() < params.iDontWantMinMessageSizeThreshold) return
+        // we need to send IDONTWANT messages to mesh peers immediately in order for them to have an effect
+        msg.topics.forEach { topic ->
```

Comment:
If there are more than 1 topic wouldn't we send duplicate `messageId` to same peers?

Reply:
Ah that is a good point, fixed in https://github.com/libp2p/jvm-libp2p/pull/374/commits/bbe42fa4f23176386325a954ad0990d0e9bd75fc